### PR TITLE
Linking sol libraries with contracts

### PR DESCRIPTION
**Pull request by yann300**

### CODE REVIEW

File: src/ClientModel.cpp [hunk 1]
```diff
@@ -592,8 +591,18 @@ void ClientModel::executeSequence(vector<TransactionSettings> const& _sequence)
 				if (transaction.functionId.isEmpty() || transaction.functionId == ctrInstance.first)
 				{
 					bytes param = encoder.encodedData();
+					bytes contractCode = m_codeModel->linkLibraries(ctrInstance.first, m_deployedLibrary);
+					eth::LinkerObject object = m_codeModel->contract(ctrInstance.first)->linkerObject();
+					if (!object.linkReferences.empty())
+					{
+						for (auto const& ref: object.linkReferences)
+							emit runFailed(QString::fromStdString(ref.second));
+						emit runFailed(ctrInstance.first + " deployment. Cannot link referenced libraries:");
```
Comment (chriseth) [inline src/ClientModel.cpp:602]: Shouldn't you return here?

File: src/ClientModel.cpp [hunk 2]
```diff
@@ -592,8 +591,18 @@ void ClientModel::executeSequence(vector<TransactionSettings> const& _sequence)
 				if (transaction.functionId.isEmpty() || transaction.functionId == ctrInstance.first)
 				{
 					bytes param = encoder.encodedData();
+					bytes contractCode = m_codeModel->linkLibraries(ctrInstance.first, m_deployedLibrary);
+					eth::LinkerObject object = m_codeModel->contract(ctrInstance.first)->linkerObject();
```
Comment (chriseth) [inline src/ClientModel.cpp:596]: Just use `bytes contractCode = object.bytecode;`?

File: src/ClientModel.h [hunk 3]
```diff
@@ -364,6 +364,7 @@ private slots:
 	QString m_dbpath;
 	ExecutionCtx m_executionCtx;
 	QString m_lastTransactionIndex;
+	QVariantMap m_deployedLibrary;
```
Comment (chriseth) [inline src/ClientModel.h]: `deployedLibraries`?

File: src/CodeModel.cpp [hunk 4]
```diff
@@ -154,9 +154,8 @@ CompiledContract::CompiledContract(const dev::solidity::CompilerStack& _compiler
 	m_contract.reset(new QContractDefinition(nullptr, &contractDefinition));
 	QQmlEngine::setObjectOwnership(m_contract.get(), QQmlEngine::CppOwnership);
 	m_contract->moveToThread(QApplication::instance()->thread());
-	eth::LinkerObject const& object = _compiler.object(_contractName.toStdString());
-	if (object.linkReferences.empty())
-		m_bytes = object.bytecode; //@todo handle unlinked objects
+	m_linkerObject = _compiler.object(_contractName.toStdString());
+	m_bytes = m_linkerObject.bytecode; // use linkLibraries() to link object
```
Comment (chriseth) [inline src/CodeModel.cpp]: Is there a good reason to have the duplication between `m_bytes` and `m_linkerObject.bytecode`?

File: src/CodeModel.h [hunk 5]
```diff
@@ -104,6 +104,11 @@ class CompiledContract: public QObject
 	QHash<LocationPair, SolidityDeclaration> const& locals() const { return m_locals; }
 	QHash<unsigned, SolidityDeclarations> const& storage() const { return m_storage; }
 
+	/// link @arg _contract to referenced libraries
+	dev::bytes linkLibraries(QVariantMap const& _deployedLibraries, QVariantMap _compiledItems);
+	/// linker object
+	eth::LinkerObject linkerObject() const { return m_linkerObject; }
```
Comment (chriseth) [inline src/CodeModel.h]: `const&`?

File: src/ClientModel.cpp (hunk 1 above)
Comment (yann300) [inline src/ClientModel.cpp:602]: there's another place in the code where runFailed is called without return. 
i did not put "return" just to stay consistent. 
But this need to review, i will do that in another pr.

**Comment by LianaHus** [comment]
@yann300 you should rebase here and resolve the conflicts

**Pull request merged**